File: app/src/main/java/de/gaffga/android/zazentimer/audio/BellCollection.java
package de.gaffga.android.zazentimer.audio;

import android.content.Context;
import android.net.Uri;
import androidx.annotation.Nullable;
import de.gaffga.android.zazentimer.Bell;
import de.gaffga.android.zazentimer.R;
import de.gaffga.android.zazentimer.bo.Section;
import java.io.File;
import java.io.FilenameFilter;
import java.util.ArrayList;

public class BellCollection {
    public static int BELL_IDX_HIGH_TONE = 0;
    public static int BELL_IDX_JAP_RHINBOWL_107 = 2;
    public static int BELL_IDX_JAP_RHINBOWL_164 = 5;
    public static int BELL_IDX_JAP_RHINBOWL_88 = 3;
    public static int BELL_IDX_JAP_RHINBOWL_90 = 4;
    public static int BELL_IDX_JAP_RHINBOWL_97 = 7;
    public static int BELL_IDX_LOW_TONE = 1;
    public static int BELL_IDX_TIB_RHINBOWL_230 = 6;
    private static final String TAG = "ZMT_BellCollection";
    private static BellCollection instance;
    private ArrayList<Bell> bells = new ArrayList<>();
    private String demoBellName;

    private BellCollection() {
    }

    public static BellCollection getInstance() {
        if (instance == null) {
            instance = new BellCollection();
        }
        return instance;
    }

    public void init(Context context) {
        this.demoBellName = context.getResources().getString(R.string.bell_name_2);
        this.bells.clear();
        this.bells.add(new Bell(getPredefinedBellUri(context, R.raw.bell1), context.getResources().getString(R.string.bell_name_1)));
        this.bells.add(new Bell(getPredefinedBellUri(context, R.raw.bell2), context.getResources().getString(R.string.bell_name_2)));
        this.bells.add(new Bell(getPredefinedBellUri(context, R.raw.dharma107), context.getResources().getString(R.string.bell_name_3)));
        this.bells.add(new Bell(getPredefinedBellUri(context, R.raw.dharmaschwarz88), context.getResources().getString(R.string.bell_name_4)));
        this.bells.add(new Bell(getPredefinedBellUri(context, R.raw.shomyo90), context.getResources().getString(R.string.bell_name_5)));
        this.bells.add(new Bell(getPredefinedBellUri(context, R.raw.tang164), context.getResources().getString(R.string.bell_name_6)));
        this.bells.add(new Bell(getPredefinedBellUri(context, R.raw.tib230), context.getResources().getString(R.string.bell_name_7)));
        this.bells.add(new Bell(getPredefinedBellUri(context, R.raw.zen97), context.getResources().getString(R.string.bell_name_8)));
        File[] listFiles = context.getFilesDir().listFiles(new FilenameFilter() { // from class: de.gaffga.android.zazentimer.audio.BellCollection.1
            @Override // java.io.FilenameFilter
            public boolean accept(File file, String str) {
                return str.startsWith("bell_");
            }
        });
        for (int i = 0; i < listFiles.length; i++) {
            this.bells.add(new Bell(getCustomBellUri(context, listFiles[i].getName()), listFiles[i].getName()));
        }
    }

    private Uri getPredefinedBellUri(Context context, int i) {
        return Uri.parse("android.resource://" + context.getPackageName() + "/" + i);
    }

    private Uri getCustomBellUri(Context context, String str) {
        return Uri.parse("file://" + context.getFilesDir() + "/" + str);
    }

    @Nullable
    public Bell getBell(String str) {
        for (int i = 0; i < this.bells.size(); i++) {
            Bell bell = this.bells.get(i);
            if (bell.getName().equals(str)) {
                return bell;
            }
        }
        return null;
    }

    public ArrayList<Bell> getBellList() {
        return this.bells;
    }

    public Bell getDemoBell() {
        return getBell(this.demoBellName);
    }

    public void release() {
        instance = null;
    }

    public Bell getBell(int i) {
        if (this.bells.size() <= i || i < 0) {
            return null;
        }
        return this.bells.get(i);
    }

    public Bell getBellForSection(Section section) {
        String str = section.bellUri;
        for (int i = 0; i < this.bells.size(); i++) {
            Bell bell = this.bells.get(i);
            if (bell.getUri().toString().equals(str)) {
                return bell;
            }
        }
        return null;
    }

    public Uri getUriForName(String str) {
        for (int i = 0; i < this.bells.size(); i++) {
            if (this.bells.get(i).getUri().toString().endsWith(str)) {
                return this.bells.get(i).getUri();
            }
        }
        return null;
    }
}

File: app/src/main/kotlin/at/priv/graf/zazentimer/audio/BellCollection.kt
package at.priv.graf.zazentimer.audio

import android.content.Context
import android.net.Uri
import androidx.annotation.Nullable
import at.priv.graf.zazentimer.R
import at.priv.graf.zazentimer.bo.Bell
import at.priv.graf.zazentimer.bo.Section
import java.io.File
import java.io.FilenameFilter

object BellCollection {
    const val BELL_IDX_HIGH_TONE: Int = 0
    const val BELL_IDX_JAP_RHINBOWL_107: Int = 2
    const val BELL_IDX_JAP_RHINBOWL_164: Int = 5
    const val BELL_IDX_JAP_RHINBOWL_88: Int = 3
    const val BELL_IDX_JAP_RHINBOWL_90: Int = 4
    const val BELL_IDX_JAP_RHINBOWL_97: Int = 7
    const val BELL_IDX_LOW_TONE: Int = 1
    const val BELL_IDX_TIB_RHINBOWL_230: Int = 6

    private const val TAG = "ZMT_BellCollection"

    private var bells: ArrayList<Bell> = ArrayList()
    private var demoBellName: String? = null

    @JvmStatic
    fun initialize(context: Context) {
        this.demoBellName = context.resources.getString(R.string.bell_name_2)
        this.bells.clear()
        this.bells.add(Bell(getPredefinedBellUri(context, R.raw.bell1), context.resources.getString(R.string.bell_name_1)))
        this.bells.add(Bell(getPredefinedBellUri(context, R.raw.bell2), context.resources.getString(R.string.bell_name_2)))
        this.bells.add(Bell(getPredefinedBellUri(context, R.raw.dharma107), context.resources.getString(R.string.bell_name_3)))
        this.bells.add(Bell(getPredefinedBellUri(context, R.raw.dharmaschwarz88), context.resources.getString(R.string.bell_name_4)))
        this.bells.add(Bell(getPredefinedBellUri(context, R.raw.shomyo90), context.resources.getString(R.string.bell_name_5)))
        this.bells.add(Bell(getPredefinedBellUri(context, R.raw.tang164), context.resources.getString(R.string.bell_name_6)))
        this.bells.add(Bell(getPredefinedBellUri(context, R.raw.tib230), context.resources.getString(R.string.bell_name_7)))
        this.bells.add(Bell(getPredefinedBellUri(context, R.raw.zen97), context.resources.getString(R.string.bell_name_8)))
        val listFiles = context.filesDir.listFiles(FilenameFilter { _: File, str: String -> str.startsWith("bell_") })
        if (listFiles != null) {
            for (file in listFiles) {
                this.bells.add(Bell(getCustomBellUri(context, file.name), file.name))
            }
        }
    }

    private fun getPredefinedBellUri(
        context: Context,
        i: Int,
    ): Uri = Uri.parse("android.resource://${context.packageName}/$i")

    private fun getCustomBellUri(
        context: Context,
        str: String,
    ): Uri = Uri.parse("file://${context.filesDir}/$str")

    @Nullable
    @JvmStatic
    fun getBell(str: String): Bell? {
        for (bell in this.bells) {
            if (bell.getName() == str) {
                return bell
            }
        }
        return null
    }

    @JvmStatic
    fun getBellList(): ArrayList<Bell> = this.bells

    @JvmStatic
    fun getDemoBell(): Bell? = demoBellName?.let { getBell(it) }

    @JvmStatic
    fun release() {
        this.bells.clear()
        this.demoBellName = null
    }

    @JvmStatic
    fun getBell(i: Int): Bell? {
        if (this.bells.size <= i || i < 0) {
            return null
        }
        return this.bells[i]
    }

    @JvmStatic
    fun getBellForSection(section: Section): Bell? {
        val str = section.bellUri
        for (bell in this.bells) {
            if (bell.uri.toString() == str) {
                return bell
            }
        }
        return null
    }

    @JvmStatic
    fun getUriForName(str: String): Uri? {
        for (bell in this.bells) {
            if (bell.uri.toString().endsWith(str)) {
                return bell.uri
            }
        }
        return null
    }
}
